### DISCUSSION TITLE
Error while running the docker image

### DISCUSSION THREAD

**Discussion by sWimmie** [serverpod/serverpod#2697] [Q&A] [answered]
I have created a staging docker image for the server but when trying to run this I get the following error:
SERVERPOD version: 2.0.2, dart: 3.5.2 (stable) (Wed Aug 28 10:01:20 2024 +0000) on "linux_x64", time: 2024-09-01 09:48:30.382044Z
mode: staging, role: monolith, logging: normal, serverId: default
Insights listening on port 8081
Server default listening on port 8080
Webserver listening on port 8082
Failed to apply database migrations. Exception: No migrations found in project.
2024-09-01 09:48:31.884502Z Internal server error. Failed to connect to Redis.
SocketException: Connection refused (OS Error: Connection refused, errno = 111), address = vkprojecten-staging.wimvandeursen.nl, port = 37010
#0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:721)
#1      _NativeSocket.connect (dart:io-patch/socket_patch.dart:984)
#2      _RawSocket.connect (dart:io-patch/socket_patch.dart:1908)
#3      RawSocket.connect (dart:io-patch/socket_patch.dart:21)
#4      Socket._connect (dart:io-patch/socket_patch.dart:2134)
#5      Socket.connect (dart:io/socket.dart:809)
#6      RedisConnection.connect (package:redis/connection.dart:21)
#7      RedisController._connect (package:serverpod/src/redis/controller.dart:75)
#8      RedisController.start (package:serverpod/src/redis/controller.dart:50)
#9      Serverpod.start.<anonymous closure> (package:serverpod/src/server/serverpod.dart:449)
<asynchronous suspension>
#10     Serverpod.start (package:serverpod/src/server/serverpod.dart:419)
<asynchronous suspension>
#11     run (package:vkprojecten_server/server.dart:101)
<asynchronous suspension>

My docker file looks like this:
FROM dart:stable AS build

WORKDIR /app
COPY . .

RUN dart pub get
RUN dart compile exe bin/main.dart -o bin/server

FROM alpine:latest

ENV runmode=staging
ENV serverid=1
ENV logging=normal
ENV role=monolith

COPY --from=build /runtime/ /
COPY --from=build /app/bin/server server
COPY --from=build /app/config/ config/
COPY --from=build /app/web/ web/

EXPOSE 8080
EXPOSE 8081
EXPOSE 8082

CMD ./server --apply-migrations --mode $runmode && bin/main --mode $runmode --server-id $serverid --logging $logging --role $role
#ENTRYPOINT ./server --apply-migrations --mode=$runmode --server-id=$serverid --logging=$logging --role=$role

So I tried to apply the migrations as well but with no result.

How to get this container running on my server?

**Comment by Isakdl** [comment, marked as answer]
So three issues here,

`CMD ./server --apply-migrations --mode $runmode && bin/main --mode $runmode --server-id $serverid --logging $logging --role $role`

`bin/main` does not exist in the docker container the binary file `server` is the entire compiled server app, you should start the server only by using this binary.

If you look at the error message it says it failed to apply the migrations, this is because you are not copying in the migration files into the container.

Add 

`COPY --from=build /app/migrations/ migrations/`

Last you are failing to connect to redis according to the logs, have you configured redis correctly?

**Reply by sWimmie** [reply]
Thanks @Isakdl I did need to copy the migrations folder. Strange that the out of the box dockerfile did not include that one.
For Redis I found that I did not have the read and write permissions set properly so that problem is solved. 
Now my server is up and running.
